recovery/elliptics_recovery/dc_server_send.py

```diff
@@ -0,0 +1,355 @@
+        index = -1
+        for index, result in enumerate(iterator, 1):
+            status = result.response.status
+            self._update_stats(start_time, index, recovers_in_progress, status)
+
+            if status < 0:
+                key = result.response.key
+                key_infos = key_infos_map[str(key)]
+                self._on_server_send_fail(status, key, key_infos, timeouted_keys, corrupted_keys, group)
+                continue
+            log.debug("Recovered key: {0}, status {1}".format(result.response.key, status))
+
+        if index < 0:
+            log.error("Server-send operation failed: group: {0}".format(group))
+            timeouted_keys = [elliptics.Id(k) for k in key_infos_map.iterkeys()]
+
+        return timeouted_keys, corrupted_keys
+
+    def _on_server_send_timeout(self, keys, key_infos_map, group):
+        '''
+        Moves keys to next bucket, if appropriate bucket meta is identical to current meta.
+        '''
+        same_meta = lambda lhs, rhs: (lhs.timestamp, lhs.size) == (rhs.timestamp, rhs.size)
+        for key in keys:
+            key_infos = key_infos_map[str(key)]
+            filtered_key_infos = self._get_unprocessed_key_infos(key_infos, group)
+            if len(filtered_key_infos) > 1:
+                current_meta = filtered_key_infos[0]
+                next_meta = filtered_key_infos[1]
+                if same_meta(current_meta, next_meta):
+                    self.buckets.on_server_send_fail(key, key_infos, next_meta.group_id)
+                    continue
+            self.result = False
+
+    def _on_server_send_fail(self, status, key, key_infos, timeouted_keys, corrupted_keys, group):
+        log.error("Failed to server-send key: {0}, group: {1}, error: {2}".format(key, group, status))
+
+        if status in (-errno.ETIMEDOUT, -errno.ENXIO):
+            timeouted_keys.append(key)
+        else:
+            if status == -errno.EILSEQ:
+                corrupted_keys.append(key)
+            next_group = self._get_next_group(key_infos, group)
+            self.buckets.on_server_send_fail(key, key_infos, next_group)
+
+    def _remove_corrupted_keys(self, keys, groups):
+        '''
+        Removes invalid keys with invalid checksum.
+        '''
+        if self.ctx.safe:
+            return
+
+        for attempt in range(self.ctx.attempts):
+            if not keys:
+                break
+
+            self.remove_session.set_groups(groups)
+
+            results = []
+            for k in keys:
+                result = self.remove_session.remove(k)
+                results.append(result)
+
+            timeouted_keys = []
+            timeouted_groups = set()
+            is_last_attempt = (attempt == self.ctx.attempts - 1)
+            for i, r in enumerate(results):
+                status = r.get()[0].status
+                log.info("Removing corrupted key: {0}, status: {1}, last attempt: {2}".format(keys[i], status, is_last_attempt))
```

Comment:
With this option, keys will not be removed.
Keys with invalid checksum will not be recovered via server-send, because they have invalid checksum.